gramineproject/gramine
# [Common,Pal,LibOS] Report crash location in debug builds

## Description of the changes 

This change adds support for determining crash location (source line, file, etc.) in situations where it cannot be known at compile time. This is done by running `addr2line` in a subprocess, and enabled (with DEBUG=1) for AddressSanitizer as well as PAL/LibOS exception handlers.

This was originally discussed in ASan pull request (#32). Unfortunately adding this to all environments (LibOS, PAL, URTS) involves some boilerplate.

## How to test this PR? 

ASan should report the crash location now:

```
$ gramine-direct run_test asan_buffer_overflow
[P1:T1:run_test] run_test("asan_buffer_overflow") ...
[P1:T1:run_test] error: asan: heap-buffer-overflow while trying to store 1 byte at 0x7ffff38c114e
[P1:T1:run_test] error: asan: the bad address is 0x7ffff38c114e (0 from beginning of access)
[P1:T1:run_test] error: asan: location: run_test_asan_buffer_overflow at shim_call.c:36, libsysdb.so+0x2bb6eb
[P1:T1:run_test] error: asan: (for a full traceback, use GDB with a breakpoint at "shim_abort")
```

You can also check LibOS, PAL, and URTS (for instance, by adding something like `volatile int* p = NULL; return *p;` somewhere):
```
[P1:T1:run_test] error: Internal memory fault at 0x00000000 (run_test_pass at shim_call.c:20,
libsysdb.so+0x2bb575, VMID = 1, TID = 1)

error: *** Unexpected memory fault occurred inside PAL (file_write at db_files.c:108,
libpal.so+0x14373, PID = 73647, TID = 73647)! ***

error: Segmentation Fault in Untrusted Code (sgx_ocall_write at sgx_enclave.c:143,
loader+0x978c)
```


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gramineproject/gramine/143)
